openmls/src/ciphersuite/mod.rs

```diff
@@ -538,6 +546,67 @@ impl Signature {
     }
 }
 
+// The size of the individual scalars. Since we only support ECDSA with P256,
+// this is 32. It would be great if evercrypt were able to return the scalar
+// size of a given curve.
+const SCALAR_SIZE: usize = 32;
+
+// The maximal overhead when DER encoding two scalars of size 32:
+// - 1 byte SEQUENCE tag
+// - 2 byte length field
+// plus two times the following header, once for each encoded scalar:
+// - 1 byte INTEGER tag
+// - 1 byte length tag
+// - 1 byte to indicate if the integer is signed
+const MAX_DER_OVERHEAD: usize = 9;
+
+impl Signature {
+    /// This function takes a DER encoded ECDSA signature and decodes it to the
+    /// bytes representing the concatenated scalars. If the decoding fails, it
+    /// will throw a `der::Error`.
+    pub(crate) fn der_decode(&self) -> Result<Vec<u8>, der::Error> {
+        // A small closure to decode a single scalar and pad it with leading
+        // zeroes until it's `SCALAR_SIZE` bytes long.
+        let decode_scalar = |decoder: &mut der::Decoder| -> Result<Vec<u8>, der::Error> {
+            let decoded_scalar = UIntBytes::decode(decoder)?;
+            let decoded_scalar_len: usize = decoded_scalar.len().try_into()?;
+            // The verification algorithm expects the scalars to be 32 bytes
+            // long, buffered with zeroes.
+            let mut padded_scalar = vec![0u8; SCALAR_SIZE - decoded_scalar_len];
+            padded_scalar.extend_from_slice(decoded_scalar.as_bytes());
+            Ok(padded_scalar)
+        };
+
+        let mut out = Vec::with_capacity(2 * SCALAR_SIZE + MAX_DER_OVERHEAD);
+        let (r, s) = der::Decoder::new(self.value.as_slice()).sequence(|decoder| {
+            let r_decoded = decode_scalar(decoder)?;
+            let s_decoded = decode_scalar(decoder)?;
+            Ok((r_decoded, s_decoded))
+        })?;
+        out.extend_from_slice(&r);
+        out.extend_from_slice(&s);
+        Ok(out)
+    }
+
+    /// This function DER encodes a given ECDSA signature consisting of bytes
+    /// representing the concatenated scalars. If the encoding fails, it will
+    /// throw a `der::Error`.
+    pub(crate) fn der_encode(raw_signature: &[u8]) -> Result<Self, der::Error> {
+        // We DER encode the ECDSA signature as per spec, assuming that
+        // `sign` returns two concatenated values (r||s).
+        let (r, s) = raw_signature.split_at(SCALAR_SIZE);
+        let r_encoded = UIntBytes::new(r)?;
+        let s_encoded = UIntBytes::new(s)?;
+        let mut bytes = [0; 2 * SCALAR_SIZE + MAX_DER_OVERHEAD];
+        let mut encoder = der::Encoder::new(&mut bytes);
+        encoder.message(&[&r_encoded, &s_encoded])?;
```

Comment:
So you don't even parse the signature? Or is the DER crate checking the OID etc. for you?

Reply:
Maybe I don't understand what's meant to go on under the hood when DER encoding an ECDSA signature, but I was under the impression that all it really is is encoding the scalars to `UIntBytes` and add the relevant DER-specific bits in the beginning as explained in the comment above `MAX_DER_OVERHEAD`. I was not aware that I needed to do any checking or parsing of the given scalars. In particular, I was mostly following what is going on [here](https://docs.rs/ecdsa/0.12.4/src/ecdsa/der.rs.html#98) since I assumed that it DER encoding didn't carry any additional security implications with it. Was I wrong in assuming that?